src/common/include/sirf/common/DataContainer.h

```diff
@@ -37,11 +37,239 @@ which rely on the same features of the items.
+		virtual float norm() const = 0;
+
+		/// calculates the sum of this container elements
+		virtual T sum() const = 0;
+
+		/// calculates the value of this container's element with the largest real part
+		virtual T max() const = 0;
+
+		/// calculates the dot product of this container with another one
+		virtual T dot(const DataContainer& dc) const = 0;
+
+		/// \c *this = the elementwise product \c x*y
+		virtual void multiply
+			(const DataContainer& x, const DataContainer& y) = 0;
+		/// \c *this = the product \c x * y with scalar y
+		virtual void multiply
+			(const DataContainer& x, const void* ptr_y) = 0;
+
+		/// \c *this = the sum \c x + y with scalar y
+		virtual void add
+			(const DataContainer& x, const void* ptr_y) = 0;
+
+		/// \c *this = the elementwise ratio \c x / y
+		virtual void divide
+			(const DataContainer& x, const DataContainer& y) = 0;
+
+		/// \c *this = the elementwise \c max(x, y)
+		virtual void maximum
+			(const DataContainer& x, const DataContainer& y) = 0;
+		virtual void maximum
+			(const DataContainer& x, const void* ptr_y) = 0;
+
+		/// \c *this = the elementwise \c min(x, y)
+		virtual void minimum
+			(const DataContainer& x, const DataContainer& y) = 0;
+		virtual void minimum
+			(const DataContainer& x, const void* ptr_y) = 0;
+
+		/// \c *this = the elementwise \c pow(x, y)
+		virtual void power
+			(const DataContainer& x, const DataContainer& y) = 0;
+		virtual void power
+			(const DataContainer& x, const void* ptr_y) = 0;
+
+		/// \c *this = the elementwise \c exp(x)
+		virtual void exp(const DataContainer& x) = 0;
+		/// \c *this = the elementwise \c log(x)
+		virtual void log(const DataContainer& x) = 0;
+		/// \c *this = the elementwise \c sqrt(x)
+		virtual void sqrt(const DataContainer& x) = 0;
+		/// \c *this = the elementwise \c sign(x)
+		virtual void sign(const DataContainer& x) = 0;
+		/// \c *this = the elementwise \c abs(x)
+		virtual void abs(const DataContainer& x) = 0;
+
+		/// \c *this = the linear combination of \c x and \c y
+		virtual void axpby(
+			const void* ptr_a, const DataContainer& x,
+			const void* ptr_b, const DataContainer& y) = 0;
+		/// alternative interface to the above
+		virtual void xapyb(
+			const DataContainer& x, const void* ptr_a,
+			const DataContainer& y, const void* ptr_b) = 0;
+
+		/// \c *this = elementwise sum of two elementwise products \c x*a and \c y*b
+		virtual void xapyb(
+			const DataContainer& x, const DataContainer& a,
+			const DataContainer& y, const DataContainer& b) = 0;
+
+		/// \c *this = elementwise sum of \c x*a and elementwise \c y*b
+		virtual void xapyb(
+			const DataContainer& a_x, const void* ptr_a,
+			const DataContainer& a_y, const DataContainer& a_b) = 0;
+
+		/// \c *this = elementwise sum of elementwise \c x*a and \c y*b
+		void xapyb(
+			const DataContainer& a_x, const DataContainer& a_a,
+			const DataContainer& a_y, const void* ptr_b)
+		{
+			xapyb(a_y, ptr_b, a_x, a_a);
+		}
+
+		static T product(T x, T y)
```

Comment:
I think that all these `static` function have to be `protected` (or even `private`). they're not part of the public interface really